truss/protostuff/protostuff.go

```diff
@@ -0,0 +1,196 @@
+		return nil, errors.Wrap(err, "unable to marshal protoc ouput to code generator request")
+	}
+
+	return req, nil
+}
+
+// ServiceFile returns the file in req that contains a service declaration.
+func ServiceFile(req *plugin.CodeGeneratorRequest, protoFileDir string) (*os.File, error) {
+	var svcFileName string
+	for _, file := range req.GetProtoFile() {
+		if len(file.GetService()) > 0 {
+			svcFileName = file.GetName()
+		}
+	}
+
+	if svcFileName == "" {
+		return nil, errors.New("passed protofiles contain no service")
+	}
+
+	svc, err := os.Open(filepath.Join(protoFileDir, svcFileName))
+
+	if err != nil {
+		return nil, errors.Wrapf(err, "could not open service file: %v\n in path: %v",
+			protoFileDir, svcFileName)
+	}
+
+	return svc, nil
+}
+
+// getProtocOutput calls exec's $ protoc with the passed protofiles and the
+// protoc-gen-truss-protocast plugin and returns the output of protoc
+func getProtocOutput(protoPaths []string) ([]byte, error) {
+	_, err := exec.LookPath("protoc-gen-truss-protocast")
+	if err != nil {
+		return nil, errors.Wrap(err, "protoc-gen-truss-protocast does not exist in $PATH")
+	}
+
+	protocOutDir, err := ioutil.TempDir("", "truss-")
+	if err != nil {
+		return nil, errors.Wrap(err, "could not create temp directory")
+	}
+	defer os.RemoveAll(protocOutDir)
+
+	err = outputGoogleImport(protocOutDir)
+	if err != nil {
+		return nil, errors.Wrapf(err, "could not write protoc imports to dir: %s", protocOutDir)
+	}
+
+	pluginCall := filepath.Join("--truss-protocast_out=", protocOutDir)
+
+	err = protoc(protoPaths, protocOutDir, pluginCall)
+	if err != nil {
+		return nil, errors.Wrap(err, "protoc failed")
+	}
+
+	fileInfo, err := ioutil.ReadDir(protocOutDir)
+	if err != nil {
+		return nil, errors.Wrapf(err, "could not read directory: %v", protocOutDir)
+	}
+
+	for _, f := range fileInfo {
+		if !f.IsDir() {
+			fPath := filepath.Join(protocOutDir, f.Name())
+			protocOut, err := ioutil.ReadFile(fPath)
+			if err != nil {
+				return nil, errors.Wrapf(err, "cannot read file: %v", fPath)
+			}
+			return protocOut, nil
+		}
+	}
+
+	return nil, errors.Errorf("no protoc output file found in: %v", protocOutDir)
+}
+
+// protoc exec's $ protoc on protoPaths
+func protoc(protoPaths []string, importDir, plugin string) error {
+	const googleAPIImportPath = "/third_party/googleapis"
+
+	cmdArgs := []string{
+		"-I" + filepath.Join(importDir, googleAPIImportPath),
+		"--proto_path=" + filepath.Dir(protoPaths[0]),
+		plugin,
+	}
+	// Append each definition file path to the end of that command args
+	cmdArgs = append(cmdArgs, protoPaths...)
+
+	protocExec := exec.Command(
+		"protoc",
+		cmdArgs...,
+	)
+
+	outBytes, err := protocExec.CombinedOutput()
+	if err != nil {
+		return errors.Wrapf(err,
+			"protoc exec failed.\nprotoc output:\n\n%v\nprotoc arguments:\n\n%v\n\n",
+			string(outBytes), protocExec.Args)
+	}
+
+	return nil
+}
+
+// mkdir acts like $ mkdir -p path
+func mkdir(path string) error {
```

Comment:
nitpick: I tend to be unfavorable torwards single line functions. I see that the filemode is being wrapped here so maybe that is case enough. Still I probably would have opted towards making the calls inline and using a package level constant. This makes it easier to see (without godef) what we are doing.